src/pybind/mgr/pg_autoscaler/module.py

```diff
@@ -291,67 +292,80 @@ def __init__(self):
-                continue
-
-            cr_name = crush_map.get_rule_by_id(p['crush_rule'])['rule_name']
-            root_id = int(crush_map.get_rule_root(cr_name))
-            pool_root[pool_name] = root_id
-            osd_count = root_map[root_id].osd_count
-            total_ideal_pg_count = root_map[root_id].pg_target
-            ideal_pg_per_pool = total_ideal_pg_count/pool_count 
-            total_pgs = ideal_pg_per_pool / p['size'] 
-            ret += nearest_power_of_two(total_pgs)
-
-        return ret
-
-    def _get_final_pg_target_and_ratio(self, capacity_ratio):
-
-        even_ratio = 1 / self._pool_count
-        used_ratio = capacity_ratio
-        
-        if used_ratio > even_ratio:
-            self._pool_count -= 1
 
-        final_ratio = max(used_ratio, even_ratio)
+    def _calc_final_pg_target(
+            self, 
+            p, 
+            pool_name, 
+            root_map,
+            root_id,
+            capacity_ratio, 
+            even_pools, 
+            bias, 
+            is_used,
+    ):
+        """
+        is_used flag used to determine if this is the first
+        pass where the caller tries to calculate/adjust pools that has 
+        used_ratio > even_ratio else this is the second pass, 
+        we calculate final_ratio by giving it 1 / pool_count 
+        of the root we are looking.
+        """
+        if is_used:
+            even_ratio = 1 / root_map[root_id].pool_count
+            used_ratio = capacity_ratio
 
-        # So what proportion of pg allowance should we be using?
-        pool_pg_target = final_ratio * self._full_pg
-        final_pg_target = nearest_power_of_two(pool_pg_target)
-        self._pgs_left -= final_pg_target
-        
-        return final_pg_target, final_ratio, pool_pg_target
+            if used_ratio > even_ratio:
+                root_map[root_id].pool_used += 1
+            else:
+                # keep track of even_pools to be used in second pass
+                # of the caller function
+                even_pools[pool_name] = p
+                return None, None, None
 
+            final_ratio = max(used_ratio, even_ratio)
+            used_pg = final_ratio * root_map[root_id].pg_target 
+            root_map[root_id].pg_left -= used_pg
+            pool_pg_target = used_pg / p['size'] * bias
 
-    def _get_pool_status(
+        else:
+            final_ratio = 1 / (root_map[root_id].pool_count - root_map[root_id].pool_used)
+            pool_pg_target = (final_ratio * root_map[root_id].pg_left) / p['size'] * bias
+
+        final_pg_target = max(p.get('options', {}).get('pg_num_min', PG_NUM_MIN),
+                nearest_power_of_two(pool_pg_target))
+
+        self.log.info("Pool '{0}' root_id {1} using {2} of space, bias {3}, "
+                      "pg target {4} quantized to {5} (current {6})".format(
+                          p['pool_name'],
+                          root_id,
+                          capacity_ratio,
```

Comment:
Here the original code prints out final_ratio instead of capacity ratio, I think that it is wrong since as we are deducting the pools, the final ratio can in some case be 1.0 if there is only 1 even pool and the rest are pools with used ratio.